main.tf

```diff
@@ -1,3 +1,62 @@
-/********************************************************************
-This file is used to implement the ROOT module.
-*********************************************************************/
+locals {
+  # tflint-ignore: terraform_unused_declarations
+  validate_kms_plan = var.kms_encryption_enabled && var.plan != "graduated-tier" ? tobool("kms encryption is only supported for graduated-tier plan") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_auth_policy = var.kms_encryption_enabled && !var.skip_iam_authorization_policy && var.existing_kms_instance_guid == null ? tobool("When var.skip_iam_authorization_policy is set to false, and var.kms_encryption_enabled to true, a value must be passed for var.existing_kms_instance_guid in order to create the auth policy.") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_kms_values = !var.kms_encryption_enabled && (var.existing_kms_instance_guid != null || var.kms_key_crn != null) ? tobool("When passing values for var.existing_kms_instance_guid or/and var.kms_key_crn, you must set var.kms_encryption_enabled to true. Otherwise unset them to use default encryption") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_kms_vars = var.kms_encryption_enabled && (var.existing_kms_instance_guid == null || var.kms_key_crn == null) ? tobool("When setting var.kms_encryption_enabled to true, a value must be passed for var.existing_kms_instance_guid and var.kms_key_crn") : true
+
+
+  # Determine what KMS service is being used for database encryption
+  kms_service = var.kms_key_crn != null ? (
+    can(regex(".*kms.*", var.kms_key_crn)) ? "kms" : (
+      can(regex(".*hs-crypto.*", var.kms_key_crn)) ? "hs-crypto" : null
+    )
+  ) : null
+}
+
+##############################################################################
+# AppID module
+##############################################################################
+
+resource "ibm_iam_authorization_policy" "policy" {
+  count                       = var.skip_iam_authorization_policy ? 0 : 1
+  source_service_name         = "appid"
+  source_resource_group_id    = var.resource_group_id
+  description                 = "Allow AppID instance to read from KMS instance"
```

Comment:
"Allow all AppID instances in the given resource group reader access to KMS instance ${var.existing_kms_instance_guid}"